### ISSUE TITLE
General Questions: Things to be careful about when new dataset used

### ISSUE THREAD

**Issue by working12** [HuguesTHOMAS/KPConv-PyTorch#180]
1. When somebody is starting with a new dataset (a bunch of point clouds), the integration can be done easily by just creating the proper `.ply` files - Is there anything else to look at when using KPConv for a new dataset?
2. How many points are expected to be there in each of the point clouds?
3. Also, what metric should one use to identify whether the model is running correctly or not? Will it be the validation set IOU scores? 
4. Also, like other deep learning models, would it be possible to overfit this model with just a single point cloud (with several points)? Trying to map the general behavior that we find in most deep learning models. (And it would be indicated by validation IoU close to 1?)
5. What should one check when the new dataset doesn't give a good performance as per the metrics mentioned above or used by you? Edit: How do you understand the number of epochs in this model? Because validation score increases pretty slowly?

**Comment by HuguesTHOMAS** [comment]
Hi @working12,

here are some answers to your questions:

>  Is there anything else to look at when using KPConv for a new dataset?

- The first answer that comes to my mind is the density of points and general size of objects you are looking for in your dataset. You have to adapt the parameters `in_radius` and `first_subsampling_dl` to your data.  There is no rule for this except common sense (don't use small input spheres if you want to classify large object like roads or buildings). Usually, some experiments (cross-validation) are required to find the best values for each dataset. Also, a very different density of points can mess up with the batch calibration, this is talked about in other issues.

> How many points are expected to be there in each of the point clouds?

- The only limit is your available RAM. We pick input spheres across the dataset so the size of each point cloud does not really matter. If your dataset is too big for your RAM, the solution can be imagined by loading only parts of the dataset every time you need new batch examples. This is not the scope of this work, which is why we do not deal with that in the repo.

> Also, what metric should one use to identify whether the model is running correctly or not? Will it be the validation set IOU scores?

 - Yes, validation IoU scores is a great tool to verify the performances (better than AP for example). I would also recommend loading the classified point clouds to see qualitative results for a deeper analysis of what happens.


> Also, like other deep learning models, would it be possible to overfit this model with just a single point cloud (with several points)? Trying to map the general behavior that we find in most deep learning models. (And it would be indicated by validation IoU close to 1?)

 - I don't understand your question. Any reference you can share? If you mean training a point cloud and validating on the same point cloud, this has very little interest in my opinion.

> What should one check when the new dataset doesn't give a good performance as per the metrics mentioned above or used by you?

 - First thing is to see the classification for yourself to try to understand what happens and if it is a bug. Then, probably the two parameters mentioned above.

> How do you understand the number of epochs in this model? Because validation score increases pretty slowly?

- We do not have a fixed list of input data like a collection of images, because we pick spheres randomly across the data (the spheres can overlap). Therefore, the epochs are defined by a parameter (`epoch_steps`) which basically says `1 epoch = N number of input spheres`. You can make bigger epochs and have convergence in fewer epochs, but this won't change the time to converge, because what ultimately matters is the number of steps performed by the optimizer. However, changing `epoch_steps` has an influence on the learning rate decay (fixed per epoch). So reducing the `epoch_steps` is equivalent to having a larger decay and thus faster convergence (be careful to not reduce too much because faster decay also means the convergence could stop before reaching the true minimum)

**Comment by working12** [comment]
@HuguesTHOMAS 
Thanks. 

> How many points are expected to be there in each of the point clouds?
>- The only limit is your available RAM. We pick input spheres across the dataset so the size of each point cloud does not really matter. If your dataset is too big for your RAM, the solution can be imagined by loading only parts of the dataset every time you need new batch examples. This is not the scope of this work, which is why we do not deal with that in the repo.

The dataset does not have many cloud samples and each cloud sample has 10^5 points (some robotics teams acquired this data for an outside world consisting of buildings, roads, waterbodies, etc). RAM is not capable of holding this entirely but I am wondering whether the model has some theoretical lower bound regarding each sample having at least 10^n points even though we would subsample it). 

>Also, like other deep learning models, would it be possible to overfit this model with just a single point cloud (with several points)? Trying to map the general behavior that we find in most deep learning models. (And it would be indicated by validation IoU close to 1?)
> - I don't understand your question. Any reference you can share? If you mean training a point cloud and validating on the same point cloud, this has very little interest in my opinion.

Like if we have a 4-layer NN and we want to check whether the model is written correctly, sometimes we are told to overfit a small data - by doing that we are sure that the model is running correctly. This type of overfitting to a small dataset to check on the model was mentioned by me as "general behavior". Here, what is a single sample? It is just a subsampled portion of the cloud scene. Now if I provide just one cloud scene with 10^5 points and the model subsamples from that only, would the model overfit?

>How do you understand the number of epochs in this model? Because validation score increases pretty slowly?

>- We do not have a fixed list of input data like a collection of images, because we pick spheres randomly across the data (the spheres can overlap). Therefore, **_the epochs are defined by a parameter (epoch_steps) which basically says 1 epoch = N number of input spheres._** You can make bigger epochs and have convergence in fewer epochs, but this won't change the time to converge, because what ultimately matters is the number of steps performed by the optimizer. However, changing epoch_steps has an influence on the learning rate decay (fixed per epoch). So reducing the epoch_steps is equivalent to having a larger decay and thus faster convergence (be careful to not reduce too much because faster decay also means the convergence could stop before reaching the true minimum)

What is controlling the number of input spheres? My guess - is it the radius and the kernel points hyperparameters? I don't see "number of input spheres" as a hyperparameter directly.

**Comment by HuguesTHOMAS** [comment]
> whether the model has some theoretical lower bound regarding each sample having at least 10^n points even though we would subsample it

I never faced any issue with a low number of points, it should be good to have fewer points, because then you can create bigger batches. Just be careful if you subsample to much you might lose details that are valuable for the classification. I would say the general rule of subs is: if you as a human can still recognise the objects in the subsampled point cloud, then the network should have enough detail to do its job.

> Now if I provide just one cloud scene with 10^5 points and the model subsamples from that only, would the model overfit?

One scene with 10^5 points is really small. The model should easily overfit that. If you want to do this, a simple way would be to copy the point cloud twice, names both cloud differently, and act as if one is the training set and the other is the validation. If you want smaller, I would not subsample, but instead, extract a smaller part of the data (extract points in a small region with a software like CloudCompare for example).

> What is controlling the number of input spheres? My guess - is it the radius and the kernel points hyperparameters? I don't see "number of input spheres" as a hyperparameter directly.

- `epoch_steps` is the number of steps per epoch (so the number of batch given as input)
- `batch_num` is the targeted number of spheres per batch. Targerted because the number of spheres per batch is variable. But overall, the average number of spheres per batch should be close to `batch_num`.


Therefore number of spheres per epoch is approximatively `epoch_steps` * `batch_num`
